### ISSUE TITLE
Context Type for wasm-pack

### ISSUE THREAD

**Issue by mgattozzi** [rustwasm/wasm-pack#123]
@data-pup came up with an interesting PoC in https://github.com/ashleygwilliams/wasm-pack/pull/122. I definitely think we should consider it as a possibility going forward, but I felt the change was a bit large and drastic to do all at once and wanted to defer it to a discussion first.

We essentially would thread a Context type to determine decisions for the program. This isn't unprecedented. For instance [`cargo`](https://github.com/rust-lang/cargo/blob/757112c39f8a5a4f6ffa0b7e29d23a56f03fb931/src/cargo/core/compiler/context/mod.rs#L65-L79) and [`rustc`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/?search=Context) use this concept as well. I also happen to use this at work extensively and this can help reduce the need for multiple extra function parameters.

I think this would be a good idea but I would love to hear thoughts because:

1) This is a big change in how we maintain and think about wasm-pack
2) We have a lot of other pressing things to get done first in terms of triage
3) If we did want to do this what would our upgrade path be? How can we make this work in multiple PRs rather than one big one so we don't let anything slip through the cracks?

CC/ @ashleygwilliams as this is a really big change and I'm not sure how this fits into your vision of wasm-pack and the code base in general

**Comment by ashleygwilliams** [comment]
in general i think this pattern is a good one that we should aim for. need to think more on how we can do it iteratively tho... 🤔 does anyone else have any thoughts?  i definitely would want to do this only after #18 because at the moment i dont super trust our tests :(

**Comment by mgattozzi** [comment]
What do you mean by iteratively? I have some thoughts on how to do this but I'm not sure what you mean here

**Comment by data-pup** [comment]
It's been a little while since I did this, but I remember the `init` command being the one that caused the most trouble. We could start by refactoring the `pack` and `publish` commands first, before thinking more about the `init` command.

Also agree with you that this should come after #18

**Comment by ashleygwilliams** [comment]
this is very out of date and the code base has changed a lot since then. i'm not against this type of refactor, but let's propose it again in a new issue with the newer codebase in mind!
